docs/dev/go-style-guide.md

```diff
@@ -0,0 +1,116 @@
+
+# Go style guide
+
+In order to keep our code looking good with lots of programmers working on it, it helps to have a "style guide", so all the code generally looks quite similar. This doesn't mean there is only one "right way" to write code, or even that this standard is better than your style.  But if we agree to a number of stylistic practices, it makes it much easier to read and modify new code. Please feel free to make suggestions if there's something you would like to add or modify.
+
+We expect all contributors to be familiar with [Effective Go](https://golang.org/doc/effective_go.html) (and it's recommended reading for all Go programmers anyways). Additionally, we generally agree with the suggestions in [Uber's style guide](https://github.com/uber-go/guide/blob/master/style.md) and use that as a starting point.
+
+## Code Structure
+
+Perhaps more key for code readability than good commenting is having the right structure. As a rule of thumb, try to write in a logical order of importance, taking a little time to think how to order and divide the code such that someone could scroll down and understand the functionality of it just as well as you do. A loose example of such order would be:
+
+- Constants, global and package-level variables.
+- Main struct definition.
+- Options (only if they are seen as critical to the struct else they should be placed in another file).
+- Initialization/start and stop of the service functions.
+- Public functions (in order of most important).
+- Private/helper functions.
+- Auxiliary structs and function (can also be above private functions or in a separate file).
+
+## General
+
+- Use `gofumpt` to format all code upon saving it (or run `make format`).
+- Think about documentation, and try to leave godoc comments, when it will help new developers.
+- Every package should have a high level doc.go file to describe the purpose of that package, its main functions, and any other relevant information.
```

Comment:
While we don't have this at the moment I think it would be nice for apps, lightclients and core submodules to have a `doc.go` with a brief description.

Reply:
Let's do that in a follow-up PR!